server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -3897,6 +3899,78 @@ public void testFlushOnIdle() throws Exception {
+        final MockLogAppender mockLogAppender = new MockLogAppender();
+        try {
+            CountDownLatch readyToCompleteFlushLatch = new CountDownLatch(1);
+            IndexShard shard = newStartedShard(false, Settings.EMPTY, config -> new InternalEngine(config) {
+                @Override
+                protected void commitIndexWriter(final IndexWriter writer, final Translog translog) throws IOException {
+                    try {
+                        readyToCompleteFlushLatch.await();
+                    } catch (InterruptedException e) {
+                        throw new AssertionError(e);
+                    }
+                    super.commitIndexWriter(writer, translog);
+                }
+            });
+
+            final int manyDocs = randomIntBetween(1, 5);
+            for (int i = 0; i < manyDocs; i++) {
+                indexDoc(shard, "_doc", Integer.toString(i));
+            }
+
+            mockLogAppender.start();
+            Loggers.addAppender(LogManager.getLogger(Engine.class), mockLogAppender);
+
+            shard.flushOnIdle(0); // flush happens in the background using the flush threadpool
+            assertFalse(shard.isActive());
+
+            // Wait for log message that flush acquired lock immediately
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "should see first flush getting lock immediately",
+                    Engine.class.getCanonicalName(),
+                    Level.TRACE,
+                    "acquired flush lock immediately"
+                )
+            );
+            assertBusy(mockLogAppender::assertAllExpectationsMatched);
+
+            // While the previous flushOnIdle request is happening, issue a second flush request with waitIfOngoing=false
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "should see second flush returning since it will not wait for the ongoing flush",
+                    Engine.class.getCanonicalName(),
+                    Level.TRACE,
+                    "detected an in-flight flush, not blocking to wait for it's completion"
+                )
+            );
+            assertFalse(shard.flush(new FlushRequest().waitIfOngoing(false).force(false)));
+            mockLogAppender.assertAllExpectationsMatched();
+
+            // Allow first flush to complete
+            readyToCompleteFlushLatch.countDown();
+
+            // Wait for first flush to log a message that it released the flush lock
+            mockLogAppender.addExpectation(
+                new MockLogAppender.SeenEventExpectation(
+                    "should see first flush releasing lock",
+                    Engine.class.getCanonicalName(),
+                    Level.TRACE,
+                    "released flush lock"
+                )
+            );
+            assertBusy(mockLogAppender::assertAllExpectationsMatched);
+
```

Comment:
Hi @henningandersen , could not do it immediately because I called flush() instead of a flushOnIdle request. Modified the test slightly to also do another flushOnIdle request to test that the active flag turns true. Thanks for the suggestion!